ansible_ai_connect/ai/api/views.py

```diff
@@ -134,11 +127,99 @@
+            model_name = model_mesh_client.get_model_id(
+                org_id, self.validated_data.get("model", "")
+            )
+            return model_name
+        except (WcaNoDefaultModelId, WcaModelIdNotFound, WcaSecretManagerError):
+            return ""
+
+    def handle_exception(self, exc):
+        self.exception = exc
+
+        # Mapping between the internal exceptions and the API exceptions (with a message and a code)
+        mapping = [
+            (WcaInvalidModelId, WcaInvalidModelIdException),
+            (WcaBadRequest, WcaBadRequestException),
+            (WcaKeyNotFound, WcaKeyNotFoundException),
+            (WcaModelIdNotFound, WcaModelIdNotFoundException),
+            (WcaNoDefaultModelId, WcaNoDefaultModelIdException),
+            (WcaEmptyResponse, WcaEmptyResponseException),
+            (WcaCloudflareRejection, WcaCloudflareRejectionException),
+            (WcaUserTrialExpired, WcaUserTrialExpiredException),
+        ]
+
+        for original_class, new_class in mapping:
+            if isinstance(exc, original_class):
+                exc = new_class(cause=exc)
+                break
+        response = super().handle_exception(exc)
+        return response
+
+    def get_ids(self):
+        allowed = ["explanationId", "generationId", "suggestionId"]
+        # Return the ids we want to include in the exception messages
+        ret = {}
+        for k, v in self.validated_data.items():
+            if k in allowed and v:
+                ret[k] = v
+            elif isinstance(v, dict):
+                for subk, subv in v.items():
+                    if subk in allowed and subv:
+                        ret[subk] = subv
+        return ret
+
+    def dispatch(self, request, *args, **kwargs):
+        start_time = time.time()
+        self.exception = False
+        self.schema1_event = None
+        response = super().dispatch(request, *args, **kwargs)
+
+        if self.schema1_event:
+            if hasattr(self.schema1_event, "duration"):
+                duration = round((time.time() - start_time) * 1000, 2)
+                self.schema1_event.duration = duration
+            self.schema1_event.modelName = self._get_model_name(request.user.org_id)
+            self.schema1_event.set_exception(self.exception)
+            # NOTE: We need to wait to store the request because keys like
+            # request._request._prompt_type are stored in the request object
+            # during the processing of the request.
+            self.schema1_event.set_request(request)  # Read the note above
+            # before moving the line ^
+
+            # NOTE: We also want to include the final response in the event, we do that
+            # we need to jump back and do it from within a final middleware that wrap
+            # everything.
+            import ansible_ai_connect.main.middleware
+
+            ansible_ai_connect.main.middleware.global_schema1_event = self.schema1_event
```

Comment:
Oeeeee... this smells bad, does it not!?!? Is this thread safe?

Reply:
Yes, this is not thread-safe, but use one process per request.

Reply:
I now use a Semaphore to protect the global variable.

Reply:
Could/should we just stuff the schema1_event in the request object instead?

Reply:
I think both strategy are ugly and can be avoided. At this stage we use the middleware only to get the final HTTP code/message, and those can easily be retrieved from the final exception that we return https://github.com/ansible/ansible-ai-connect-service/pull/1147/files#diff-ecfb6919dfd8379aafba96af7457b253e4dce528897dfe6bfc207ca2b3b2ada9R169

Django will return a 500 unless the exception has a specific code. There is still potential 500 if the exception happens outside of the view, and we can use the `exception_handler()` to catch them.